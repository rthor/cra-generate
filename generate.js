'use strict'

const EOL = require('os').EOL
const fs = require('fs')
const mkdir = require('mkpath')
const chalk = require('chalk')
const path = require('path')
const nodetree = require('nodetree')
const caseTransform = require('./lib/case-transform')

function template(file) {
  const filePath = path.resolve(__dirname, 'templates/component/', file)
  return {
    filePath,
    content: fs.readFileSync(filePath, 'utf8'),
  }
}

function replaceVars(content, componentName, fileName, cssExtension, semi) {
  return content.replace(/\$Name\$/g, componentName)
                .replace(/\$name\$/g, fileName)
                .replace(/\$semi\$/g, semi ? ';' : '')
                .replace(/\$css-ext\$/g, cssExtension)
}

function saveToFile(filePath, content) {
  fs.writeFileSync(filePath, content, 'utf8')
}

function relativePath(src) {
  return `./${path.relative(process.cwd(), src)}`
}

function getComponentPath(componentName, directory, fileName) {
  const cwd = process.cwd()
  const root = path.join(cwd, 'src')
  mkdir.sync(root)

  const dir = path.join(root, directory)
  mkdir.sync(dir)
  const componentPath = path.join(dir, fileName)
  const exists = fs.existsSync(componentPath)

  if (exists) {
    console.error(chalk.red(`Component ${chalk.bold(componentName)} already exists at ${relativePath(componentPath)}`))
    return process.exit(1)
  }

  mkdir.sync(componentPath)
  return componentPath
}

function implementTypeChecking(typeSystem, content) {
  let mutatedContent

  switch (typeSystem) {
    case 'flow':
      mutatedContent = `// @flow${EOL}${EOL}${content}`
      break
    default:
      mutatedContent = content
      break
  }

  return {
    content: mutatedContent,
  }
}

module.exports = function generate(component, options) {
  const fileName = caseTransform(component, options.fileFormat)
  const componentName = caseTransform(component, options.componentFormat)
  const componentPath = getComponentPath(componentName, options.directory, fileName)

  const scriptFiles = [
    template('index.js'),
    template(options.isFunctional ? 'stateless.js' : 'stateful.js'),
  ]

  if (options.test === 'jest') {
    scriptFiles.push(template('jest.js'))
  }
  const styleFiles = [template('styles.css')]
  const cssExtension = options.cssExtension.replace(/^\./, '')

  scriptFiles.forEach(script => {
    const name = path.parse(script.filePath).name
    const content = implementTypeChecking(options.typeCheck, script.content).content
    const scriptName = (
      name === 'index' ? 'index' : name === 'jest' ? `${fileName}.test` : `${fileName}`
    )
    const filePath = path.join(componentPath, `${scriptName}.js`)
    saveToFile(filePath, replaceVars(content, componentName, scriptName, cssExtension, options.semi))
  })

  styleFiles.forEach(style => {
    const filePath = path.join(componentPath, `${fileName}.${cssExtension}`)
    saveToFile(filePath, replaceVars(style.content, componentName, fileName, cssExtension, options.semi))
  })

  console.log(chalk.green(`Generated ${chalk.cyan.bold(componentName)}:\n`))

  nodetree(componentPath, {
    prune: true,
    noreport: true
  })
}
