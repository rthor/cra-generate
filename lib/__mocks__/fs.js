'use strict'

const fs = jest.genMockFromModule('fs')

fs.readFileSync = (path) => `Content at: [${path}]`

fs.existsSync = () => true

fs.writeFileSync = function(filePath, content, encoding) {
  console.log(filePath, ':')
  console.log(content)
}

module.exports = fs
