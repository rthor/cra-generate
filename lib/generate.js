'use strict'

const caseTransform = require('./case-transform')
const createFormatter = require('./formatter')
const getConfig = require('../config')
const io = require('./io')

module.exports = function generate(component, options = getConfig({})) {
  const fileName = caseTransform(component, options.fileFormat)
  const componentName = caseTransform(component, options.componentFormat)
  const componentPath = io.getComponentPath(componentName, options.directory, fileName)

  const fileFormatter = createFormatter({
    fileName,
    componentName,
    componentPath,
    semicolon: options.semi,
    typeCheck: options.typeCheck,
    cssExtension: options.cssExtension.replace(/^\./, ''),
  })

  const files = [
    { type: 'script', template: io.getTemplate('index.js') },
    { type: 'script', template: io.getTemplate(options.isFunctional ? 'stateless.js' : 'stateful.js') },
    { type: 'script', template: options.test === 'jest' ? io.getTemplate('jest.js') : null },
    { type: 'style', template: io.getTemplate('styles.css') }
  ]
    .filter(file => file.template !== null)
    .map(file => fileFormatter(file))

  for (const file of files) {
    io.writeToDisk(file.path, file.content)
  }

  return { componentName, componentPath, files }
}
