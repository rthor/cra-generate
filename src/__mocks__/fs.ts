interface fs_t {
  readFileSync: (path: string) => string
  existsSync: (path: string) => boolean
  writeFileSync: () => void
  __setupMockDirectories: (directories: Array<string>) => void
}

interface foldersMap_t {
  [directory: string]: boolean
}

const fs: fs_t = jest.genMockFromModule("fs")
let foldersThatExists: foldersMap_t = {}

fs.readFileSync = path => `Content at: [${path}]`

fs.existsSync = path => Boolean(path in foldersThatExists)

fs.writeFileSync = () => {}

fs.__setupMockDirectories = directories => {
  foldersThatExists = Object.create(null)
  for (let index = 0; index < directories.length; index++) {
    foldersThatExists[directories[index]] = true
  }
}

export = fs
