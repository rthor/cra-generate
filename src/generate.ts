import caseTransform = require("./case-transform")
import getConfig = require("./config")
import * as templates from "../templates"
import * as io from "./io"

export = function generate(component: string, options: any = {}) {
  options = getConfig(options)

  const fileName = caseTransform(component, options.fileFormat)
  const componentName = caseTransform(component, options.componentFormat)
  const componentPath = io.getComponentPath(
    componentName,
    options.directory,
    fileName
  )

  const files = templates(options.typeCheck || "")({
    fileName,
    componentName,
    componentPath,
    noTest: !options.test,
    isFunctional: options.isFunctional,
    semiColon: options.semi ? ";" : "",
    cssExtension: options.cssExtension.replace(/^\./, ""),
  })

  for (const file of files) {
    io.writeToDisk(file.filePath, file.content)
  }

  return { componentName, componentPath, files }
}
