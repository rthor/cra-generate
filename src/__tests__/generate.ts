jest.mock("fs")
jest.mock("mkpath")

import generate = require("../generate")
import * as path from "path"

describe("generate", () => {
  it("should fail if component already exists", () => {
    require("fs").__setupMockDirectories([
      path.resolve(process.cwd(), "src", "components", "Test"),
    ])
    expect(() => {
      generate("test")
    }).toThrowErrorMatchingSnapshot()
  })

  it("should generate a component with default config", () => {
    const data = generate("test-component")
    expect(data.componentName).toMatchSnapshot()
    expect(data.files.map(({ content }) => content)).toMatchSnapshot()
  })

  it("should generate functional components", () => {
    const data: any = generate("test-component", { functional: true })
    const file = data.files.find(
      ({ fileName }) => fileName === "TestComponent.js"
    )
    expect(file.content).toMatchSnapshot()
  })

  it("should add semi colons", () => {
    const data: any = generate("test-component", { semi: true })
    expect(data.files.map(({ content }) => content)).toMatchSnapshot()
  })

  it("should change css extension", () => {
    const data: any = generate("t", { cssExtension: "sass" })
    expect(
      data.files.find(({ fileName }) => fileName === "T.sass").content
    ).toMatchSnapshot()
  })

  it("should skip test file", () => {
    const data = generate("t", { test: "none" })
    expect(data.files.length).toBe(3)
  })

  it("should use flow type", () => {
    const data: any = generate("t", { typeCheck: "flow" })
    const scripts = data.files
      .filter(({ fileName }) => fileName.includes(".js", -1))
      .map(({ content }) => content)
    expect(scripts).toMatchSnapshot()
  })

  it("should use customized directory", () => {
    const data = generate("t", { directory: "foo" })
    expect(data.componentPath).toContain("/src/foo/T")
  })
})
