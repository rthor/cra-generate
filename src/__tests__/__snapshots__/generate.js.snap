// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate should add semi colons 1`] = `
Array [
  ".TestComponent {}
",
  "export { default } from './TestComponent';
",
  "import React from 'react';
import ReactDOM from 'react-dom';
import TestComponent from './TestComponent';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TestComponent />, div);
});
",
  "import React, { Component } from 'react';
import './TestComponent.css';

class TestComponent extends Component {
  state = {};

  render() {
    return (
      <div className=\\"TestComponent\\"></div>
    );
  }
}

export default TestComponent;
",
]
`;

exports[`generate should change css extension 1`] = `
".T {}
"
`;

exports[`generate should fail if component already exists 1`] = `"Component Test already exists at ./src/components/Test"`;

exports[`generate should generate a component with default config 1`] = `"TestComponent"`;

exports[`generate should generate a component with default config 2`] = `
Array [
  ".TestComponent {}
",
  "export { default } from './TestComponent';
",
  "import React from 'react';
import ReactDOM from 'react-dom';
import TestComponent from './TestComponent';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TestComponent />, div);
});
",
  "import React, { Component } from 'react';
import './TestComponent.css';

class TestComponent extends Component {
  state = {};

  render() {
    return (
      <div className=\\"TestComponent\\"></div>
    );
  }
}

export default TestComponent;
",
]
`;

exports[`generate should generate functional components 1`] = `
"import React from 'react';
import './TestComponent.css';

const TestComponent = ({}) => (
  <div className=\\"TestComponent\\"></div>
);

export default TestComponent;
"
`;

exports[`generate should use flow type 1`] = `
Array [
  "// @flow

export { default } from './T';
",
  "// @flow

import React from 'react';
import ReactDOM from 'react-dom';
import T from './T';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<T />, div);
});
",
  "// @flow

import React, { Component } from 'react';
import './T.css';

class T extends Component {
  state = {};

  render() {
    return (
      <div className=\\"T\\"></div>
    );
  }
}

export default T;
",
]
`;

exports[`generate should use typescript 1`] = `
Array [
  "export { default } from './T';
",
  "import * as React from 'react';
import * as ReactDOM from 'react-dom';
import T from './T';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<T />, div);
});
",
  "import * as React from 'react';
import './T.css';

class T extends React.Component<{}, {}> {
  public state = {};

  public render() {
    return (
      <div className=\\"T\\" />
    );
  }
}

export default T;
",
]
`;
